### ISSUE TITLE
Discussion about Atom-community

### ISSUE THREAD

**Issue by aminya** [atom-ide-community/atom#4]
I am pleased to say that we now have a parallel fork of Atom, which plans to speed up the process of development and improvements. We will release the builds under the `atom-community` name. The contributions will eventually become available in the upstream `atom` builds.
https://github.com/atom-ide-community/atom

You can check the projects to see more details about the plan: https://github.com/atom-ide-community/atom/projects 
In this repository, we use projects as our backlog and a way to track the changes we make. This will help us allowing the enhancements to be considered for the upstream repository. For example, we plan to include atom-ide packages to our `atom-community` builds by default.

https://github.com/atom-ide-community/atom/projects

**Comment by DeeDeeG** [comment]
There are a couple of ways we can handle patches.

- Treat our work as a series of patches or changes that we continually rebase on upstream `master`
- Have long-lived branches here, which we merge upstream `master` into continually

(And for completeness sake: if we ever needed to hard fork, we could stop tracking upstream entirely. But I'd really like to avoid doing that, because upstream are our friends in my opinion, and it "reduces the team size" drastically if we have to try to keep Atom running on our own/cut upstream out of the process. And all that without the expertise of long-standing Atom maintainers.)

I'm not sure I strongly prefer either approach.

The long-lived "branches" approach makes us a bit friendlier to contributors here, but we are going to probably run into some really gnarly merges over time.

If we keep our repo rebased on upstream `master` it makes it easy to focus on upstreaming our work. At the cost of history at our fork being a bit nonsensical, or at least hard to follow. Since we're changing the commit history all the time with rebases. This also raises the question of how/if we keep older states of our fork around. If we use branches/tags for this, it will produce lots of branches/tags and be cluttered (hard to sort through). If we don't the work history for this project is very short-lived, so current maintainers need to be very careful how they merge/rebase stuff.

---

Short summary: Having long-lived branches here leaves a legible history, but keeping the history legible comes at the cost of potentially difficult merges going forward. Rebasing constantly is simpler for us, but makes it messier or even impossible to look back at this fork's past work.

If anyone has thoughts on this, I'd be interested to hear it.

**Comment by DeeDeeG** [comment]
Having said all that, I'm leaning toward long-lived branches that we merge upstream `master` into. Keeps the history readable, and I think we can handle the merge conflicts... rebasing isn't immune to merge conflicts either.

**Comment by DeeDeeG** [comment]
Off-topic: It's getting late where I am, so I'm going to sleep. Will take a look at all this tomorrow if I get the chance.

**Comment by aminya** [comment]
Thank you for the answers. I will reply, once we get more people in.

**Comment by aminya** [comment]
Added the pull bot to make pull requests from upstream. We should use rebase as long as there are no conflicts. I will need to approve the pull requests it makes, so a human review will happen.

**Comment by DeeDeeG** [comment]
@lkashef @darangi if you have thoughts about this fork, I am also all ears/interested to know what you would have to say.

---

@aminya I think we can't necessarily expect our pull requests to be gotten to in any particular order. Also, our commit history will have strange stuff like merging their upstream `master` into our `master` -- if we open a PR from our `master`, they would have to accept all of the strange merge commits and essentially this is like them becoming a downstream of _us_. I don't think it makes much sense unless they "squash and merge" our PRs, which is something I haven't seen them doing.

I think we should give upstream just the commits containing the changes for the PR we are proposing, directly on top of their `master` branch, and then accept whatever their commit history is into ours by merging (their `master` --> out `upstream_master` --> merge into our `master`).

Open to discussing it more. Just that the underlying `git` software needs to be capable of it, whatever it is.

**Comment by aminya** [comment]
> @lkashef @darangi if you have thoughts about this fork, I am also all ears/interested to know what you would have to say.
> 
> @aminya I think we can't necessarily expect our pull requests to be gotten to in any particular order. Also, our commit history will have strange stuff like merging their upstream `master` into our `master` -- if we open a PR from our `master`, they would have to accept all of the strange merge commits and essentially this is like them becoming a downstream of _us_. I don't think it makes much sense unless they "squash and merge" our PRs, which is something I haven't seen them doing.
> 

Draft pull requests cannot be merged until marked as ready. We should only mark the first one in the sequence, and that is the only one they can merge. 


> I think we should give upstream just the commits containing the changes for the PR we are proposing, directly on top of their `master` branch, and then accept whatever their commit history is into ours by merging (their `master` --> out `upstream_master` --> merge into our `master`).
> 
> Open to discussing it more. Just that the underlying `git` software needs to be capable of it, whatever it is.

The commit history is not important here. The point is to improve the code itself. Although rebasing will keep the history clean.

**Comment by DeeDeeG** [comment]
I'd like to point out something: we can't post more than one PR from our `master` branch at the same time -- All PRs from the same branch will point to the same commit (the one at the tip).

I think your plan is technically sound, if reviewers at upstream review our PRs promptly and merge them with little/no delay... which considering human factors and the recent pace of the upstream repo is sort of unrealistic in my view. And it verges on bad contributor/maintainer ettiquette to presume our faster rate of development/PRs will be welcome. It is still a maintenance burden to review our PRs, regardless of whether the code contributions are a welcome sight or not.

I don't think we can assume ahead of time that any of our work will be upstreamed, period. Much less in a timely manner. NONE of this is to disparage upstream maintainers, who have my utmost respect. We should have realistic expectations, and a flexible workflow, if we are going to work somewhat independently of upstream but still try to communicate back.

It's easy to just go our own way, at our own speed, but that implies a harder fork than I've been picturing. Short of a hard fork, or a more one-way relationship where we take code we like from upstream and add stuff on top, we need to bake "waiting for review" into our process.

**Comment by aminya** [comment]
> I'd like to point out something: we can't post more than one PR from our `master` branch at the same time -- All PRs from the same branch will point to the same commit (the one at the tip).
> 

We will make separate branches. That's how we want to release! See the upstream branches, each one is a version:
https://github.com/atom/atom/branches We will do the same. One pull request for each release.

**Comment by DeeDeeG** [comment]
I there we can have stuff that only we want, such as the pullbot config, without posting it in our PRs to upstream?

**Comment by aminya** [comment]
I will remove the pullbot config. It is saved in the bot when committed for the first time. If there were more things like that, we need to write some code to automate making pull requests instead of doing things the regular way, but I do not think this situation will arise.

**Comment by aminya** [comment]
I have updated many of the atom/npm packages that are in the atom organization or widely used by atom packages. We should also integrate those into our atom builds. I should register them under a new name for doing that.

A list of those that I remember now:
atom/packages:
- status-bar https://github.com/atom/status-bar/pull/236
- line ending selector https://github.com/atom/atom/pull/20542
- tree-view (tests remaining): https://github.com/atom/tree-view/pull/1341

npm:
- underscore-plus: https://github.com/atom/underscore-plus/pull/21
- atom-select-list https://github.com/atom/atom-select-list/pull/28
- space-pen: https://github.com/aminya/space-pen
- atom-space-pen-views https://github.com/atom/atom-space-pen-views/pull/23
- etch (WIP): https://github.com/atom/etch/pull/82

We also need to integrate useful community packages as I mentioned in https://github.com/atom-ide-community/atom/projects/4

**Comment by aminya** [comment]
@DeeDeeG Are you able to edit the projects in https://github.com/atom-ide-community/atom/projects?

**Comment by DeeDeeG** [comment]
No.

I would need to be given some more permissions in this repo's settings, or in the atom-ide-community org settings.

**Comment by DeeDeeG** [comment]
Things I can't do: Edit other people's comments/PRs, open/close other people's issues/PRs, modify anything in the atomcommunity Azure DevOps organization, probably some other stuff.

**Comment by aminya** [comment]
This should be fixed now.

**Comment by DeeDeeG** [comment]
I'm busy enough with getting this repo started, and I easily lose focus, so that I'm reluctant to modify projects and other meta stuff outside of issues/PRs. Might change if we have larger tasks that are clearly structured, but it's all a little on-the-fly/ad-hoc until things get more situated.

I do like to backtrack and document things after the fact, but a roadmap is a lot to ask for IMO this early on. We are just paving the paths of how we do this thing at the moment.

**Comment by aminya** [comment]
Yes, once we close #1, we can actually start releasing.

**Comment by DeeDeeG** [comment]
I have a bunch of permissions around the repo now, thanks. :heart:

**Comment by DeeDeeG** [comment]
Moving off-topic conversation from #1 to here.

From https://github.com/atom-ide-community/atom/issues/1#issuecomment-656511353:

> the control is in the hands of the community and not the upstream team, which now only has few developers

Yes, that I agree with. It seems you are right. We need freedom, or else, why fork?

A compromise: I will do my best to maintain clean branches of my work (or any work here that upstream probably wants) at my fork. That can be my personal project.

Here at this fork, I will not protest anymore if there are radical changes. Some consideration of "keeping rebases or merges of upstream back to us easy" is prudent and forward-looking, but another reasonable priority can be "not getting stuck." I'll follow your lead on the direction this takes, honestly.

I do have some PRs I did upstream that i'd like to merge here, though. Mostly https://github.com/atom/atom/pull/20856.

**Comment by DeeDeeG** [comment]
I am going to personally look to stop working on CI in favor of features and stuff in the core app.

My priorities for CI:
- Simple config (keeps the maintainers of the config from getting confused)
- Improvements to CI should be upstreamable, or we should be aware up-front that we are committing to maintaining changes if/when they need fixing)
- Reliable.
- Doesn't do anything weird that we didn't expect or take up too much of the project's resources.

Upstream's CI laregely checks those boxes for me. I just want to get upstream's CI config working here and move on to working on features.

**Comment by aminya** [comment]
> I am going to personally look to stop working on CI in favor of features and stuff in the core app.

Yes. That is what is important. I will finish CI changes after we merge our open PRs. I do not want us to spend more time on the CI. We have done most of the things that could help us.

Other than those that merged already, we have 5 left:
- CI-only improvements:
	#46 increases reusability of the steps and makes them parallelizable.
	#33 increases the chance of using the cache even if we modify package.json.
	#43 cache npm cache
- Test improvements:
	#31 Speeds up tests (windows and macOS) by a huge amount.
- Using latest build facilities:
	#44 uses the latest Windows image for building Atom
- Easing development:
	#10 allows starting atom development right away

After these finished, I will put my focus on [bundling packages with atom-community](https://github.com/atom-ide-community/atom/projects/4). For the first release, we should at least integrate one package to make `atom-community` exciting for the users. Saying that we have improved CI does not make any difference to the end-user. I plan to integrate x-terminal and then minimap-plus as our first packages. If you see other repositories on this organization, I have made improvements in those too.

**Comment by DeeDeeG** [comment]
> they are interested in the changes.

At the time Mazen commented, the title of that PR was "Parallelize Windows CI", so I intepreted that comment as narrowly being interested in taking the tests that had been running in one `job` on Windows, and splitting them up into multiple `job`s, jut like it's done on macOS.

Upstream is more conservative than us on infrastructure changes, especially from outside contributors. So I had assumed they would have wanted that in whatever minimal-changes way it could be sent upstream as a patch. Without asking for clarification, I suppose I can't say for sure.

Heavy use of templates isn't required in order to split up the tests, for example. So they presumably don't want to accept that _and_ parallelization in one PR.

> Upstream CI is not reliable

And neither is ours. The same tests must pass here as there, and they time out. We cannot fix that, as far as I know, without finding a way to ensure individual tests can complete faster. (Which requires troubleshooting slowness in the testing process overall. I also get timeouts locally, not just in CI.)

Though it would be nice for the test run/test `job` to exit instantly when any required test fails.

> They have to run a CI 11 times until it passes.

That was an extreme outlier case. It's in no way representative of the norm for upstream. They usually have to rerun one, two, or perhaps three times. Four or five in a bad case. 11 was a total outlier.

And so would we, if our tests timed out 11 times. Paralellization of Windows x64 reduces the chances of this significantly. So that's good. But I don't expect it will make CI fail infrequently at this fork. Just much faster to finish tests in three parallel `job`s, rather than serially in one `job`, on a good clean run. And also much, much faster to finish the "Re-run failed jobs" follow-up runs. That's good, but let's keep this benefit in perspective. Our CI will need re-runs still, even with this change.

(I don't know if you noticed, but I have been manually re-running failing builds at our fork, to make sure we have fundamentally working code and CI still. But I don't mean to hide the fact that our CI needs re-runs.)

**Comment by DeeDeeG** [comment]
Should we start adding release notes, even for developer-facing changes or CI changes?

Our target audience is honestly those who want to tinker more deeply with Atom and its infrastructure at the moment. Maybe these changes are relevant to our target demographic. Just a thought.

**Comment by aminya** [comment]
What do you mean by release notes? For the final releases? If so, we will do it.

**Comment by DeeDeeG** [comment]
I mean the last bit of all the PR templates asks for Release Notes. If we turn "auto draft releases" on, then I think those are scraped and added to the draft release for you.

**Comment by bennypowers** [comment]
This is really encouraging! Thank you for taking this job on.

There are some outstanding but low-hanging issues with `ide-typescript` which could give a lot of added value with little effort if they were maintained:

- https://github.com/atom/ide-typescript/pull/156
- https://github.com/atom/ide-typescript/pull/160

What do you think is the outlook on maintaining `ide-typescript` or a fork under `atom-ide-community`?

Thank you again for breathing new life into my favourite editor.

**Comment by aminya** [comment]
@bennypowers Hi Benny! Why don't you use `atom-typescirpt`? Actually you can use [`atom-ide-javascript`](https://atom.io/packages/atom-ide-javascript) which sets up typescript as well.

We will look into those PRs, but the packages I mentioned are up to date and under active development.

**Comment by jeff-hykin** [comment]
> Treat our work as a series of patches or changes that we continually rebase on upstream master
Have long-lived branches here, which we merge upstream master into continually

Was there ever a resolution to the merging strategy?  @DeeDeeG

**Comment by DeeDeeG** [comment]
@jeff-hykin This was discussed here: https://github.com/atom-ide-community/atom/issues/28, and we came to essentially the same conclusion.

Practically speaking: We have been using plain merges (the kind that generates a merge commit -- no rebase or squash) to bring in updates from upstream `master` branch lately, and plan to continue to do so.

**Comment by aminya** [comment]
> Speaking from (moderate) experience, I think continually merging in upstream master is the way to go.

Exactly. That's what we found. Rebasing (any force-push) makes a lot of problems for our pull requests.

**Comment by jeff-hykin** [comment]
> @jeff-hykin This was discussed here: #28, and we came to essentially the same conclusion.

Great to hear! Sorry I looked for the conclusion but didn't see one.

**Comment by jeff-hykin** [comment]
Also just as an FYI, I'm here because I'm interested parallel work: creating a refresh of Atom. I've used VS Code for many years and contributed a decent bit (I maintain their C++ syntax), but after exploring the VS Codebase and seeing how unmaintainable it is, I've switched back to Atom for good.

A standard IDE interface would be part of the refresh, which is why I'm mentioning this here. I'm curious if you guys are opposed or interested in a re-branded Atom (not an extension, but also not a hard fork).

The main goal is standardization; How can an IDE debugger use a terminal, when it doesn't know which terminal(s) you have? How can a theme customize terminal-colors when there is no standard for a terminals (and no standard terminal)? Even beyond extension-interactions, I think the lack of standardization is killing community maintenance. Looking at terminals themselves; there's platformio (a fork), teminus (a fork of that), `terminal-tab` using Xterm.js (a fork of term.js used by platformio), and the most up-to-date is a fork of `terminal-tab`: `iv-terminal`. Yet platformio is the most common. The community can't update/maintain `iv-terminal` if they don't even know it exists, and new coder's won't know to pick it over platformio.

However, part of the refresh is UI layout changes, different default settings, extensions, and a reduction of node-gyp dependencies.

**Comment by DeeDeeG** [comment]
That sounds beyond my skill level, personally.  I am invested in making Atom's infrastructure sustainable, and updating dependencies, mostly. Even when that takes substantial care, updating uses of dependencies' APIs, substantial re-writes of sections of code, whatever it takes. I'm interested in keeping Atom modern and just "alive" and current. (Part of my calculation is that upstream is still interested in working on features, but a lot of their work hours are caught up in basic maintenance. I want to ease the burden for maintenance tasks and free up work hours for features and other concerns.) I'm self-taught and not an expert, but I can figure out a fair amount of Atom if I get an evening or two to stare at the code and do research, and read the PRs and past discussion that led to the current code and designs.

That said, part of the work we're doing is to pave the bumps in the road for forks of Atom. A project is healthier when downstream contributors can fork you, run the infrastructure that supports you, and make changes/propose them back to upstream when needed.

If upstream hasn't merged one of our PRs yet aiming to improve the forking experience, we will have the answers for any of the problems we've solved. (For example, we have CI 100% working, if you count remaining open PRs to fix Release Branches and Nightly). Ideally we will have those manual interventions needed written as documentation, and where possible stuff should "just work" at upstream and forks with no intervention, and we will do that if and where there's a readily apparent way to do so. But also, we're so far quite happy to answer the volume of questions we've been getting, which has been low in these early days.

None of what we're doing is proprietary, from the code to the ethos. There is plenty of room for more than one, and even many players in this space. Where there is something of interest to this fork and another fork, I know I'd be interested in that, and out of encouraging the health of the ecosystem, I'd like to put effort into working together. Whether that needs to happen in this repo or another one is IMO kind of down to something more specific than you've mentioned, feature by feature.

I might lean against changing the UI, personally, or adding a lot of features to core, but I think I'm the more conservative about changes among the two of us who have been mainly working on this fork at the moment. And if it can be justified and discussed to come to a consensus, I can be swayed on things. But more specifics would be good before I would want to say yea or nay.

tl;dr I'm open to anything, but there's only so far my expertise will go and so far I can think to go with features and drastic changes before I get a bit skeptical or concerned about it. But I would encourage and support efforts to fork upstream, us, or use code from both upstream and us as a base for new features. We still aim to coordinate with upstream where possible, so I think the right person to ask about a given piece of tech can be looped in to the discussion. If not, hopefully the state we're leaving the ecosystem in is vital enough, and pleasant enough to work on top of, so forking is a fruitful endeavor despite us/without having to ask us anything :P.

**Comment by aminya** [comment]
> Also just as an FYI, I'm here because I'm interested parallel work: creating a refresh of Atom. I've used VS Code for many years and contributed a decent bit (I maintain their C++ syntax), but after exploring the VS Codebase and seeing how unmaintainable it is, I've switched back to Atom for good.
> 
> A standard IDE interface would be part of the refresh, which is why I'm mentioning this here. I'm curious if you guys are opposed or interested in a re-branded Atom (not an extension, but also **not** a hard fork).
> 
> The main goal is standardization; How can an IDE debugger use a terminal, when it doesn't know which terminal(s) you have? How can a theme customize terminal-colors when there is no standard for a terminals (and no standard terminal)? Even beyond extension-interactions, I think the lack of standardization is killing community maintenance. Looking at terminals themselves; there's platformio (a fork), teminus (a fork of that), `terminal-tab` using Xterm.js (a fork of term.js used by platformio), and the most up-to-date is a fork of `terminal-tab`: `iv-terminal`. Yet platformio is the most common. The community can't update/maintain `iv-terminal` if they don't even know it exists, and new coder's won't know to pick it over platformio.
> 
> However, part of the refresh is UI layout changes, different default settings & extensions, and a reduction of node-gyp dependencies.

Thanks for the interest. VsCode is taking a closed approach in their codebase and extensions. They don't allow any extension to add any UI which keeps the extensions mostly simple. However, Atom gives the real power of improving the environment to the developers. That's why I think Atom is the superior choice for me.

See these issues. People are seeking simple freedom for adding toolbar, float windows, etc. The things that can be added "very simply" in Atom.
https://github.com/microsoft/vscode/issues/41309
https://github.com/microsoft/vscode/issues/10121

We think we can easily cover the VsCode API by adding wrapper functions, and we have started the `atom-vscode` package. Since VsCode does not allow adding any UI, having such a compatibility layer will allow any extension to work without adjusting much of the code!
https://github.com/atom-ide-community/atom-vscode

There are plans for standardization. It is undergoing in atom-ide-language-client and atom-ide-base. 

I have started with improving the old packages. We can achieve so much just by optimizing and modernizing the JavaScript packages. See this list for example https://github.com/atom-ide-community/atom/issues/4#issuecomment-653903375. 

In addition, for the hot parts of the code that we cannot optimize in JavaScript, I plan to use AssemblyScript or Rust to convert them to WASM. I will upload some template packages later.

Regarding the modernization of the native modules, replacing `@atom/nsfw` with `nsfw 2` is in my todo lists (#76 ). I think there are so many modern native modules that we can use right away. We just need to update the API a bit.

Multi-threading is my other big thing. Many things can run in parallel without affecting the main GUI thread. #74 

By the way, regarding Terminals, [X-terminal](https://github.com/bus-stop/x-terminal) is the Atom's best terminal right now and I plan to include in our Atom builds.

**Comment by jeff-hykin** [comment]
Oh I 200% agree with you @aminya about VS Code, I hate their philosophy and I have never stared the repo because of it. When I say standards I only mean defaults or "correct"  ways, but fully maintaining the proactively-hackable nature of Atom.


> We think we can cover the VS Code API

That would be a game changer! I hadn't even considered that. Sadly, I don't think it will be easy though, the API is quite large. Some actions use tools that don't exist in Atom by default. But it is doable and a large number of them can be done quite easily.

That's great to hear about Javascript package updates, I've been planning on doing the same.

I 100% agree with the move to WASM. I was actually the first person to get the Tree Sitter running in VS Code with WASM 😅. I want to compile the fuzzy-native to wasm and bundle all the language packs into one repo. 


> By the way, regarding Terminals, X-Terminal 

Thanks! I missed that one.

**Comment by jeff-hykin** [comment]
Also good to hear from you @DeeDeeG Being skeptical is fair, and being open is more than enough for me. Atoms UI style is great IMO and I want to keep that, but some things (like URI handling having an entire tab for 1 setting) don't make much sense at all.

**Comment by DeeDeeG** [comment]
Just for context, the `atom-ide-community` org here on GitHub has been around for a while, significantly longer than this forked copy of Atom core. There are a lot of passionate and talented people who have been active with that. @aminya and I recently decided to fork atom core and were invited to have access to do so under this org.

I hope that's stating it accurately. Point being, this whole thing is bigger than `atom-ide-community/atom`. I am active at the forked Atom core (this repo) but others have been digging deeper into the code and writing/re-writing extensions, IDE-related features, etc. for some time now, so they know more about all that than I do.

Whatever happens, I personally want to see that Atom core is in good health and active.

**Comment by aminya** [comment]
Closing this issue in favor of the brand-new feature of Github, "discussions".
https://github.com/atom-ide-community/atom/discussions/267
